src/minecraft/kotlin/xyz/wagyourtail/unimined/minecraft/patch/forge/AccessTransformerMinecraftTransformer.kt

```diff
@@ -317,4 +328,240 @@ object AccessTransformerMinecraftTransformer {
         }
         return output
     }
+
+    fun at2aw(at: Path, output: Path, namespace: String, mappings: MappingTreeView): Path {
+        val inReader = at.bufferedReader(
+            StandardCharsets.UTF_8,
+            1024,
+            StandardOpenOption.READ
+        )
+
+        val reader = transformFromLegacyTransformer(inReader)
+        val temp = at.resolveSibling(at.name + ".temp")
+        val tempWriter = temp.bufferedWriter(
+            StandardCharsets.UTF_8,
+            1024,
+            StandardOpenOption.CREATE,
+            StandardOpenOption.TRUNCATE_EXISTING
+        )
+        reader.copyTo(tempWriter)
+
+        val masterList = AccessTransformerList()
+        masterList.loadFromPath(temp, at.name)
+        temp.deleteIfExists()
+        val awWriter = AccessWidenerWriter()
+        awWriter.visitHeader(namespace)
+
+        val targetAccess = AccessTransformer::class.java.getField("targetAccess")
+        val targetFinalState = AccessTransformer::class.java.getField("targetFinalState")
+        val targetMember = AccessTransformer::class.java.getField("memberTarget")
+
+        val namespaceId = mappings.getNamespaceId(namespace)
+
+        masterList.accessTransformers.forEach { (_, transformers) ->
+            transformers.forEach { it ->
+                val tAccess = targetAccess.get(it) as Modifier
+                val tFinal = targetFinalState.get(it) as AccessTransformer.FinalState
+                val member = targetMember.get(it) as Target<*>
+
+                when (member.type) {
+                    TargetType.FIELD -> {
+                        val fieldName = member.targetName()
+
+                        val fieldDesc = mappings.getClass(member.className, namespaceId).fields.first { fieldView ->
```

Comment:
mapping trees aren't always complete. only tiny is guarenteed to be complete. also, why do we need mappings in the first pace

Reply:
We need mappings because At doesn't include field descriptor and Aw requires having it, so it's a way to get it.

The mappings are also needed to convert At wildcard as there is no such thing in Aw.